File: core/runtime/types/chat_types.py
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, TypeAlias

from core.runtime.storage import JSONValue

if TYPE_CHECKING:
    from core.runtime.types.import_export_types import ExportArtifactRequest, ExportArtifactResult
    from core.runtime.types.skill_types import SkillExecutionRequest, SkillExecutionResult
    from core.runtime.types.workspace_types import ServiceMutationRequest, ServiceMutationResult

JSONObject: TypeAlias = dict[str, JSONValue]


@dataclass(frozen=True, slots=True)
class ChatMessageSnapshot:
    message_state_id: str
    chat_message_id: str
    chat_role: str
    linked_object_id: str | None
    linked_revision_id: str | None
    payload: JSONObject


@dataclass(frozen=True, slots=True)
class ChatSessionSnapshot:
    session_id: str
    project_id: str
    novel_id: str | None
    title: str | None
    runtime_origin: str
    created_by: str
    messages: tuple[ChatMessageSnapshot, ...]


@dataclass(frozen=True, slots=True)
class OpenChatSessionRequest:
    project_id: str
    created_by: str
    runtime_origin: str
    novel_id: str | None = None
    title: str | None = None
    source_ref: str | None = None


@dataclass(frozen=True, slots=True)
class OpenChatSessionResult:
    session_id: str
    project_id: str
    novel_id: str | None
    title: str | None
    runtime_origin: str


@dataclass(frozen=True, slots=True)
class GetChatSessionRequest:
    session_id: str


@dataclass(frozen=True, slots=True)
class ChatMessageRequest:
    chat_message_id: str
    chat_role: str
    payload: JSONObject


@dataclass(frozen=True, slots=True)
class ChatTurnRequest:
    project_id: str
    created_by: str
    runtime_origin: str
    user_message: ChatMessageRequest
    assistant_message: ChatMessageRequest
    session_id: str | None = None
    novel_id: str | None = None
    title: str | None = None
    source_ref: str | None = None
    workbench_type: str | None = None
    source_object_id: str | None = None
    source_revision_id: str | None = None
    mutation_requests: tuple[ServiceMutationRequest, ...] = ()
    export_requests: tuple[ExportArtifactRequest, ...] = ()
    skill_requests: tuple[SkillExecutionRequest, ...] = ()


@dataclass(frozen=True, slots=True)
class ChatTurnResult:
    session_id: str
    user_message_state_id: str
    assistant_message_state_id: str
    assistant_content: str
    mutation_results: tuple[ServiceMutationResult, ...]
    export_results: tuple[ExportArtifactResult, ...]
    skill_results: tuple[SkillExecutionResult, ...]

File: packages/shared/src/types/chat.types.ts
import type { JSONObject } from "./json.js";

// ─── Chat Messages ────────────────────────────────────────────────────────────

export interface ChatMessageRequest {
  chat_message_id: string;
  chat_role: string;
  payload: JSONObject;
}

export interface ChatMessageSnapshot {
  message_state_id: string;
  chat_message_id: string;
  chat_role: string;
  linked_object_id: string | null;
  linked_revision_id: string | null;
  payload: JSONObject;
}

export interface ChatSessionSnapshot {
  session_id: string;
  project_id: string;
  novel_id: string | null;
  title: string | null;
  runtime_origin: string;
  created_by: string;
  messages: ChatMessageSnapshot[];
}

// ─── Chat Requests / Results ──────────────────────────────────────────────────

export interface OpenChatSessionRequest {
  project_id: string;
  created_by: string;
  runtime_origin: string;
  novel_id?: string;
  title?: string;
  source_ref?: string;
}

export interface OpenChatSessionResult {
  session_id: string;
  project_id: string;
  created_by: string;
  runtime_origin: string;
  novel_id: string | null;
  title: string | null;
  source_ref: string | null;
}

export interface GetChatSessionRequest {
  session_id: string;
}

export interface ChatTurnRequest {
  session_id?: string;
  project_id: string;
  novel_id?: string;
  title?: string;
  workbench_type?: string;
  user_message: ChatMessageRequest;
  assistant_message: ChatMessageRequest;
  source_object_id?: string;
  source_revision_id?: string;
  mutation_requests?: unknown[];
  export_requests?: unknown[];
  skill_requests?: unknown[];
}

export interface ChatTurnResult {
  session_id: string;
  user_message_state_id: string;
  assistant_message_state_id: string;
  assistant_content: string;
  mutation_results?: unknown[];
  export_results?: unknown[];
  skill_results?: unknown[];
}
